src/Guard.php

```diff
@@ -350,125 +335,114 @@ protected function removeFromStorage($name)
+        return $this->appendTokenToRequest($request, $token);
     }
 
     /**
-     * Setter for persistentTokenMode
+     * @param ServerRequestInterface $request
+     * @param $pair
      *
-     * @param bool $persistentTokenMode True to use the same token throughout the session (unless there is a validation error),
-     * false to get a new token with each request.
-     * @return $this
+     * @return ServerRequestInterface
      */
-    public function setPersistentTokenMode($persistentTokenMode)
+    protected function appendTokenToRequest(ServerRequestInterface $request, array $pair): ServerRequestInterface
     {
-        $this->persistentTokenMode = $persistentTokenMode;
-        return $this;
+        $name = $this->prefix . '_name';
+        $value = $this->prefix . '_value';
+        return $request
+            ->withAttribute($name, $pair[$name])
+            ->withAttribute($value, $pair[$value]);
     }
 
     /**
-     * Setter for storageLimit
+     * @param ServerRequestInterface  $request
+     * @param RequestHandlerInterface $handler
      *
-     * @param integer $storageLimit Value to set
-     * @return $this
-     */
-    public function setStorageLimit($storageLimit)
-    {
-        $this->storageLimit = (int)$storageLimit;
-        return $this;
-    }
-
-    /**
-     * Getter for persistentTokenMode
+     * @return ResponseInterface
      *
-     * @return bool
+     * @throws Exception
      */
-    public function getPersistentTokenMode()
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        return $this->persistentTokenMode;
-    }
+        if (in_array($request->getMethod(), ['POST', 'PUT', 'DELETE', 'PATCH'])) {
+            $body = $request->getParsedBody();
+            $name = null;
+            $value = null;
 
-    /**
-     * @return string
-     */
-    public function getTokenName()
-    {
-        return isset($this->keyPair[$this->getTokenNameKey()]) ? $this->keyPair[$this->getTokenNameKey()] : null;
+            if (is_array($body)) {
+                $name = $body[$this->prefix . '_name'] ?? null;
+                $value = $body[$this->prefix . '_value'] ?? null;
+            }
+
+            if (!$this->validateToken($name, $value)) {
+                if (!$this->persistentTokenMode && is_string($name)) {
+                    $this->removeTokenFromStorage($name);
+                }
+
+                $request = $this->appendNewTokenToRequest($request);
+                return $this->handleFailure($request, $handler);
+            }
+        }
+
+        if (!$this->persistentTokenMode || !$this->loadLastKeyPair()) {
+            $request = $this->appendNewTokenToRequest($request);
+        } elseif ($this->persistentTokenMode) {
+            $pair = $this->loadLastKeyPair() ? $this->keyPair : $this->generateToken();
+            $request = $this->appendTokenToRequest($request, $pair);
```

Comment:
`$pair` of method `Slim\Csrf\Guard::appendTokenToRequest()` expects `array`, but here `$pair` is `array|null` because `\Slim\Csrf\Guard::$keyPair` is type hinted as so.

Reply:
so `$this->loadLastKeyPair()` returns a `true` bool if `$this->keyPair` is not `null`. It could never be `null`, if that was the case then `generateToken()` would be called.

Reply:
You are absolutely correct. Although I doubt that phpstan would detect that - in case you would like to add phpstan later on.